firrtl/src/main/scala/firrtl/ir/Serializer.scala

```diff
@@ -234,6 +241,25 @@ object Serializer {
             indent += 1
           case LayerBlockEnd =>
             indent -= 1
+          case ContractBegin(info, names, exprs) =>
+            doIndent()
+            b ++= "contract"
+            if (names.nonEmpty) {
+              b ++= " "
+              names.zipWithIndex.foreach { case (name, idx) =>
+                if (idx > 0) b ++= ", "
+                b ++= name
+              }
```

Comment:
Maybe try the following:

```suggestion
              b ++= names.mkString(" ", ", ", "")
```

Ditto for the expressions.

Scala's `mkString` is basically `llvm::interleave*`. It has two forms: a one-argument form which interleaves the argument and a three argument form which has a "start", "interleave", and "end".

There's a slight risk to this as it does create the string before passing it to the `StringBuilder`. However, I think that's probably okay here given that these are small. (I would not advise using `mkString` for interleaving newlines between modules, e.g.)

Reply:
Very nice, I wasn't aware `mkString` 👍 I don't think the same thing will work for expressions though, since they call `s` to implicitly populate `b`, so there's no string for the expression to intercept 😢